sympy/polys/tests/test_python_powerseriesring.py

```diff
@@ -0,0 +1,120 @@
+from sympy import QQ, ZZ
+from sympy.polys.python_powerseriesring import (PythonPowerSeriesRingZZ,
+                                                PythonPowerSeriesRingQQ)
+from sympy.testing.pytest import raises
+
+def test_PowerSeriesRingZZ():
+    R = PythonPowerSeriesRingZZ(5)
+    assert isinstance(R, PythonPowerSeriesRingZZ)
+    assert R.domain == ZZ
+    assert R.prec == 5
+    assert R.one == ([1], None)
+    assert R.zero == ([0], None)
+    assert R.gen == ([0, 1], None)
+
+def test_PowerSeriesRingQQ():
+    R = PythonPowerSeriesRingQQ()
+    assert isinstance(R, PythonPowerSeriesRingQQ)
+    assert R.domain == QQ
+    assert R.prec == 6
+    assert R.one == ([QQ(1)], None)
+    assert R.zero == ([QQ(0)], None)
+    assert R.gen == ([QQ(0), QQ(1)], None)
+
+# Series for testing
+R_ZZ = PythonPowerSeriesRingZZ(5)
+R_QQ = PythonPowerSeriesRingQQ(10)
+
+def test_negative():
+    assert R_ZZ.negative(([1, -2, -3], 4)) == ([-1, 2, 3], 4)
+    assert R_QQ.negative(([QQ(1, 2), QQ(3, 4)], 6)) == ([QQ(-1, 2), QQ(-3, 4)], 6)
+
+def test_equal():
+    assert R_ZZ.equal(([1, 2, 3], 4), ([1, 2, 3], 4)) is True
+    assert R_ZZ.equal(([1, 2, 3], 4), ([1, 2], None)) is False
+    assert R_QQ.equal(([QQ(1, 2), QQ(3, 4)], 6), ([QQ(1, 2), QQ(3, 4)], None)) is False
+    assert R_QQ.equal(([QQ(1, 2), QQ(3, 4)], None), ([QQ(1, 2)], None)) is False
+
+def test_add():
+
+    QQDUP1 = [QQ(1, 2), QQ(3, 4), QQ(1, 2), QQ(7, 8), QQ(2, 1)]
+    QQDUP2 = [QQ(1, 3), QQ(2, 5), QQ(3, 4)]
+
+    # Both are series
+    assert R_ZZ.add(([1, 2, 3], 4), ([4, 5], 3)) == ([5, 7, 3], 3)
+    assert R_QQ.add((QQDUP1, 6), (QQDUP2, 5)) == ([QQ(5, 6), QQ(23, 20),
+        QQ(5, 4), QQ(7, 8), QQ(2, 1)], 5)
+
+    # One is DUP and one series
+    assert R_ZZ.add(([7, 14, 21, 28], 4), ([0, 7], None)) == ([7, 21, 21, 28], 4)
+    assert R_QQ.add((QQDUP1, 5), (QQDUP2, None)) == ([QQ(5, 6), QQ(23, 20),
+        QQ(5,4), QQ(7, 8), QQ(2, 1)], 5)
+
+    # Both DUPs
+    assert R_ZZ.add(([1, 3, 1, 2, 2], None), ([10, 20], None)) == ([11, 23, 1,
+                                                                    2, 2], 5)
+    assert R_QQ.add((QQDUP1, None), (QQDUP2, None)) == ([QQ(5, 6), QQ(23, 20),
+        QQ(5, 4), QQ(7, 8), QQ(2, 1)], None)
+
+def test_subtract():
+
+    QQDUP1 = [QQ(3, 5), QQ(2, 3), QQ(4, 7), QQ(5, 9), QQ(1, 2)]
+    QQDUP2 = [QQ(1, 4), QQ(1, 6), QQ(2, 5)]
+    QQDUP3 = [QQ(5, 3), QQ(-2, 7), QQ(3, 11), QQ(7, 13)]
+
+    # Both are series
+    assert R_ZZ.subtract(([5, 8, 6], 4), ([2, 3], 3)) == ([3, 5, 6], 3)
+    assert R_QQ.subtract((QQDUP1, 6), (QQDUP2, 5)) == ([QQ(7, 20), QQ(1, 2),
+        QQ(6, 35), QQ(5, 9), QQ(1, 2)], 5)
```

Comment:
As much as possible we should want these tests to be generic over the different implementations like:
```python
for R in [R_ZZ, R_QQ, R_FLINT_ZZ, R_FLINT_QQ]:
    # tests involving R
```
This can also be done with pytest's parametrised tests.

Reply:
We don't want to write duplicate tests all the time for each of the 4 different series ring types. Especially in future there will be many more than 4 so we want to have a good test infrastructure that can test many implementations.